File: modules/expo-bluesky-swiss-army/ios/Visibility/VisibilityViewManager.swift
import Foundation

class VisibilityViewManager {
  static let shared = VisibilityViewManager()

  private let views = NSHashTable<VisibilityView>(options: .weakMemory)
  private var currentlyActiveView: VisibilityView?
  private var screenHeight: CGFloat = UIScreen.main.bounds.height

  func addView(_ view: VisibilityView) {
    self.views.add(view)
  }

  func removeView(_ view: VisibilityView) {
    self.views.remove(view)
  }

  func updateActiveView() {
    DispatchQueue.main.async {
      var activeView: VisibilityView?

      if self.views.count == 1 {
        let view = self.views.allObjects[0]
        if view.isViewableEnough() {
          activeView = view
        }
      } else if self.views.count > 1 {
        let views = self.views.allObjects
        var mostVisibleView: VisibilityView?
        var mostVisiblePosition: CGRect?

        views.forEach { view in
          if !view.isViewableEnough() {
            return
          }

          guard let position = view.getPositionOnScreen() else {
            return
          }

          if position.minY >= 150 {
            if mostVisiblePosition == nil {
              mostVisiblePosition = position
            }

            if let unwrapped = mostVisiblePosition,
               position.minY <= unwrapped.minY {
              mostVisibleView = view
              mostVisiblePosition = position
            }
          }
        }

        activeView = mostVisibleView
      }

      if activeView == self.currentlyActiveView {
        return
      }

      self.clearActiveView()
      if let view = activeView {
        self.setActiveView(view)
      }
    }
  }

  private func clearActiveView() {
    if let currentlyActiveView = self.currentlyActiveView {
      currentlyActiveView.setIsCurrentlyActive(isActive: false)
      self.currentlyActiveView = nil
    }
  }

  private func setActiveView(_ view: VisibilityView) {
    view.setIsCurrentlyActive(isActive: true)
    self.currentlyActiveView = view
  }
}

File: modules/expo-bluesky-swiss-army/android/src/main/java/expo/modules/blueskyswissarmy/visibilityview/VisibilityViewManager.kt
package expo.modules.blueskyswissarmy.visibilityview

import android.graphics.Rect

class VisibilityViewManager {
  companion object {
    private val views = HashMap<Int, VisibilityView>()
    private var currentlyActiveView: VisibilityView? = null
    private var prevCount = 0

    fun addView(view: VisibilityView) {
      this.views[view.id] = view

      if (this.prevCount == 0) {
        this.updateActiveView()
      }
      this.prevCount = this.views.count()
    }

    fun removeView(view: VisibilityView) {
      this.views.remove(view.id)
      this.prevCount = this.views.count()
    }

    fun updateActiveView() {
      var activeView: VisibilityView? = null
      val count = this.views.count()

      if (count == 1) {
        val view = this.views.values.first()
        if (view.isViewableEnough()) {
          activeView = view
        }
      } else if (count > 1) {
        val views = this.views.values
        var mostVisibleView: VisibilityView? = null
        var mostVisiblePosition: Rect? = null

        views.forEach { view ->
          if (!view.isViewableEnough()) {
            return
          }

          val position = view.getPositionOnScreen() ?: return@forEach
          val topY = position.centerY() - (position.height() / 2)

          if (topY >= 150) {
            if (mostVisiblePosition == null) {
              mostVisiblePosition = position
            }

            if (position.centerY() <= mostVisiblePosition!!.centerY()) {
              mostVisibleView = view
              mostVisiblePosition = position
            }
          }
        }

        activeView = mostVisibleView
      }

      if (activeView == this.currentlyActiveView) {
        return
      }

      this.clearActiveView()
      if (activeView != null) {
        this.setActiveView(activeView)
      }
    }

    private fun clearActiveView() {
      this.currentlyActiveView?.setIsCurrentlyActive(false)
      this.currentlyActiveView = null
    }

    private fun setActiveView(view: VisibilityView) {
      view.setIsCurrentlyActive(true)
      this.currentlyActiveView = view
    }
  }
}
